README.md
# scaph2cc

A small utility to filter and convert ⚡ [Scaphandre](https://github.com/hubblo-org/scaphandre) json exporter output to a format suitable for 🌳 CarbonCrush API.

## Usage

Invoke scaph2CC through command line. It expects the following parameters:

- the scaphandre json analysis file as input
- the name of the file to be generated
- the process name that you want to use to filter the input file
- additional carbon crush metadata:
  - application id
  - ci branch measured
  - url of the current ci pipeline
  - commit id

Scaph2CC returns a json summary file that can be ingested by CarbonCrush API and a junit test report.

```bash
cargo run -- -i tests/scaphandre-full-report.json -o tests/carboncrush-data.json --junit-report-file tests/carboncrush-report.xml --app-id myapp-123 --branch master --ci-pipeline-url "http://ci.com/master/123456" --process-name stress-ng --commit-sha d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33
```

```bash
scaph2cc 0.0.5
Olivier de Meringo <demeringo@gmail.com>
The arguments of scap2cc CLi

USAGE:
    scaph2cc [OPTIONS] --input-file <INPUT_FILE> --output-file <OUTPUT_FILE> --process-name <PROCESS_NAME> --app-id <APP_ID> --branch <BRANCH> --ci-pipeline-url <CI_PIPELINE_URL> --commit-sha <COMMIT_SHA>

OPTIONS:
    -a, --app-id <APP_ID>
            Carbon crush app id

    -b, --branch <BRANCH>
            Name of the current branch

    -c, --commit-sha <COMMIT_SHA>
            commit identifier (sha)

    -h, --help
            Print help information

    -i, --input-file <INPUT_FILE>
            Name of the scaphandre result file to use as input

    -j, --junit-report-file <JUNIT_REPORT_FILE>
            Name of the junit report to generate [default: carboncrush-report.xml]

    -o, --output-file <OUTPUT_FILE>
            Name of the carbon crush file to generate

    -p, --process-name <PROCESS_NAME>
            The name of the process to filter

    -u, --ci-pipeline-url <CI_PIPELINE_URL>
            ci pipeline URL

    -V, --version
```

### Carbon Crush input Format

```json
{
  "consumption": "7269278.142857143",
  "duration" : "10.00",
  "appId": "123",
  "branch": "master",
  "ciPipelineUrl": "http://ci.com/master/123456",
  "commitSha":"d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33",
  "energy" : "72692781.42857143"
}
```
## Usage in CI

scaph2cc is included in the [scaphandre-node-ci docker image](https://github.com/demeringo/scaphandre-node-ci/)

## Thanks

🌳 _All credits go to the Scaphandre project and its great communitity_: <https://github.com/hubblo-org/scaphandre>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased] - ReleaseDate

## [0.1.1] - 2023-08-08

- Add an option to pass pid on command line (not implemented yet).

## [0.1.0] - 2022-06-17

### Added

- A new `energyWatthours` field is added to the generated json result. It contains the energy value in Watt-hours. The previously existing field (`energy`) is kept but should not be used because its unit is unpractical (microWatts-second).

## [0.0.5] - 2022-06-08

### Added

- Generate a junit report (xml format). Adds a `--junit-report-file` parameter to pass the name of the report to generate (without parameter, default to `carboncrush-report.xml`)

## [0.0.4] - 2022-04-20

- first working version

<!-- Examples below -->
<!--
## [1.0.0] - 2017-06-20
### Added
- New visual identity by [@tylerfortune8](https://github.com/tylerfortune8).
- Version navigation.
- Links to latest released version in previous versions.
- "Why keep a changelog?" section.
- "Who needs a changelog?" section.
- "How do I make a changelog?" section.
- "Frequently Asked Questions" section.
- New "Guiding Principles" sub-section to "How do I make a changelog?".
- Simplified and Traditional Chinese translations from [@tianshuo](https://github.com/tianshuo).
- German translation from [@mpbzh](https://github.com/mpbzh) & [@Art4](https://github.com/Art4).
- Italian translation from [@azkidenz](https://github.com/azkidenz).
- Swedish translation from [@magol](https://github.com/magol).
- Turkish translation from [@karalamalar](https://github.com/karalamalar).
- French translation from [@zapashcanon](https://github.com/zapashcanon).
- Brazilian Portugese translation from [@Webysther](https://github.com/Webysther).
- Polish translation from [@amielucha](https://github.com/amielucha) & [@m-aciek](https://github.com/m-aciek).
- Russian translation from [@aishek](https://github.com/aishek).
- Czech translation from [@h4vry](https://github.com/h4vry).
- Slovak translation from [@jkostolansky](https://github.com/jkostolansky).
- Korean translation from [@pierceh89](https://github.com/pierceh89).
- Croatian translation from [@porx](https://github.com/porx).
- Persian translation from [@Hameds](https://github.com/Hameds).
- Ukrainian translation from [@osadchyi-s](https://github.com/osadchyi-s).

### Changed
- Start using "changelog" over "change log" since it's the common usage.
- Start versioning based on the current English version at 0.3.0 to help
translation authors keep things up-to-date.
- Rewrite "What makes unicorns cry?" section.
- Rewrite "Ignoring Deprecations" sub-section to clarify the ideal
  scenario.
- Improve "Commit log diffs" sub-section to further argument against
  them.
- Merge "Why can’t people just use a git log diff?" with "Commit log
  diffs"
- Fix typos in Simplified Chinese and Traditional Chinese translations.
- Fix typos in Brazilian Portuguese translation.
- Fix typos in Turkish translation.
- Fix typos in Czech translation.
- Fix typos in Swedish translation.
- Improve phrasing in French translation.
- Fix phrasing and spelling in German translation.

### Removed
- Section about "changelog" vs "CHANGELOG".

## [0.3.0] - 2015-12-03
### Added
- RU translation from [@aishek](https://github.com/aishek).
- pt-BR translation from [@tallesl](https://github.com/tallesl).
- es-ES translation from [@ZeliosAriex](https://github.com/ZeliosAriex).

## [0.2.0] - 2015-10-06
### Changed
- Remove exclusionary mentions of "open source" since this project can
benefit both "open" and "closed" source projects equally.

## [0.1.0] - 2015-10-06
### Added
- Answer "Should you ever rewrite a change log?".

### Changed
- Improve argument against commit logs.
- Start following [SemVer](https://semver.org) properly.

## [0.0.8] - 2015-02-17
### Changed
- Update year to match in every README example.
- Reluctantly stop making fun of Brits only, since most of the world
  writes dates in a strange way.

### Fixed
- Fix typos in recent README changes.
- Update outdated unreleased diff link.

## [0.0.7] - 2015-02-16
### Added
- Link, and make it obvious that date format is ISO 8601.

### Changed
- Clarified the section on "Is there a standard change log format?".

### Fixed
- Fix Markdown links to tag comparison URL with footnote-style links.

## [0.0.6] - 2014-12-12
### Added
- README section on "yanked" releases.

## [0.0.5] - 2014-08-09
### Added
- Markdown links to version tags on release headings.
- Unreleased section to gather unreleased changes and encourage note
keeping prior to releases.

## [0.0.4] - 2014-08-09
### Added
- Better explanation of the difference between the file ("CHANGELOG")
and its function "the change log".

### Changed
- Refer to a "change log" instead of a "CHANGELOG" throughout the site
to differentiate between the file and the purpose of the file — the
logging of changes.

### Removed
- Remove empty sections from CHANGELOG, they occupy too much space and
create too much noise in the file. People will have to assume that the
missing sections were intentionally left out because they contained no
notable changes.

## [0.0.3] - 2014-08-09
### Added
- "Why should I care?" section mentioning The Changelog podcast.

## [0.0.2] - 2014-07-10
### Added
- Explanation of the recommended reverse chronological release ordering.

## [0.0.1] - 2014-05-31
### Added
- This CHANGELOG file to hopefully serve as an evolving example of a
  standardized open source project CHANGELOG.
- CNAME file to enable GitHub Pages custom domain
- README now contains answers to common questions about CHANGELOGs
- Good examples and basic guidelines, including proper date formatting.
- Counter-examples: "What makes unicorns cry?"

[Unreleased]: https://github.com/olivierlacan/keep-a-changelog/compare/v1.0.0...HEAD

src/main.rs
//! # scaph2cc
//!
//! `scaph2cc` a small CLI that allows converting Scaphandre output to CarbonCrush API input format.
//!
//! `scaph2cc` filters the Scaphandre JSON output to aggregate measures on specific processes
//! and add some context information (extracted from CLI options) about the current build
//! context.  

mod carboncrush_exporter;
use std::path::PathBuf;

use crate::carboncrush_exporter::*;

mod scaphandre_reader;

use clap::Parser;

/// The arguments of scaph2cc CLi
#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]
struct Args {
    /// Name of the scaphandre result file to use as input
    #[clap(short, long, parse(from_os_str))]
    input_file: PathBuf,

    /// Name of the carbon crush file to generate
    #[clap(short, long, parse(from_os_str))]
    output_file: PathBuf,

    /// Name of the junit report to generate
    #[clap(
        short,
        long,
        parse(from_os_str),
        default_value = "carboncrush-report.xml"
    )]
    junit_report_file: PathBuf,

    /// The name of the process to filter
    #[clap(short = 'n', long)]
    process_name: String,

      /// The pid of the process to filter
      #[clap(short = 'p', long)]
      process_id: String,

    /// Carbon crush app id
    #[clap(short, long)]
    app_id: String,

    /// Name of the current branch
    #[clap(short, long)]
    branch: String,

    /// ci pipeline URL
    #[clap(short = 'u', long)]
    ci_pipeline_url: String,

    /// commit identifier (sha)
    #[clap(short = 'c', long)]
    commit_sha: String,
}

/// scaph2cc CLI, used to filter scaphandre data and upload it to CarbonCrush API.
fn main() {
    let args = Args::parse();
    let app_id = args.app_id.as_str();
    let branch = args.branch.as_str();
    let pipeline_url = args.ci_pipeline_url.as_str();
    let scaphandre_json_file = args.input_file;
    let carboncrush_json_file = args.output_file;
    let junit_report = args.junit_report_file;
    let process_name = args.process_name.as_str();
    let commit_sha = args.commit_sha.as_str();

    let average_power_microwatts =
        scaphandre_reader::average_consumption_microwatt(&scaphandre_json_file, process_name);
    let duration_s =
        scaphandre_reader::process_duration_seconds(&scaphandre_json_file, process_name);

    let carbon_crush_result = build_carboncrush_result(
        average_power_microwatts,
        app_id,
        branch,
        commit_sha,
        pipeline_url,
        duration_s,
    );

    println!(
        "Done. Average consumption (uW): {}, Duration (s): {}, Total energy (uWs): {}, Total energy (Wh): {}",
        average_power_microwatts, duration_s, carbon_crush_result.energy, carbon_crush_result.energy_watthours
    );

    save_carboncrush_file(&carbon_crush_result, carboncrush_json_file);

    save_as_junit_report(&carbon_crush_result, junit_report);
}


src/cc_format.rs
use serde::Deserialize;
use serde::Serialize;
use std::fs::File;
use std::path::Path;

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CarbonCrushResult {
    pub consumption: String,
    pub app_id: String,
    pub duration: String,
    pub branch: String,
    pub commit_sha: String,
    pub energy: String,
    pub ci_pipeline_url: String,
}

pub fn read_cc_file(filename: &str) -> CarbonCrushResult {
    let json_file_path = Path::new(filename);
    let file = File::open(json_file_path).expect("file not found");
    let result: CarbonCrushResult = serde_json::from_reader(file).expect("error while reading");
    return result;
}

pub fn print_cc_file(c: CarbonCrushResult) {
    println!("Full results {:?}", c);
    println!(
        "Extract: appid:{}  pipeline url:{} consumption:{}, energy {}, duration",
        c.app_id, c.ci_pipeline_url, c.consumption, c.duration
    )
}

pub fn save_cc_file(carbon_crush_result: CarbonCrushResult, filename: &str) {
    println!("Saving results: {:?} to {}", carbon_crush_result, filename);
    // Save the JSON structure into the other file.
    std::fs::write(
        filename,
        serde_json::to_string_pretty(&carbon_crush_result).unwrap(),
    )
    .unwrap();
}

pub fn build_cc_result(
    consumption: f32,
    app_id: &str,
    branch: &str,
    commit_sha: &str,
    ci_pipeline_url: &str,
    energy: f32,
    duration: f32,
) -> CarbonCrushResult {
    CarbonCrushResult {
        consumption: consumption.to_string(),
        app_id: app_id.to_string(),
        energy: energy.to_string(),
        branch: branch.to_string(),
        ci_pipeline_url: ci_pipeline_url.to_string(),
        commit_sha: commit_sha.to_string(),
        duration: duration.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_reading_carbon_crush_results() {
        read_cc_file("./tests/carbon-crush-sample.json");
    }
    #[test]
    fn test_print_cc_file() {
        print_cc_file(read_cc_file("./tests/carbon-crush-sample.json"));
    }

    #[test]
    fn test_save_cc_file() {
        let filename = "test-generated-result.json";
        let carbon_crush_result = build_cc_result(
            123.0,
            "myapp1",
            "main",
            "d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33",
            "http://whatever/job/123",
            1230.0,
            10.0,
        );
        save_cc_file(carbon_crush_result, filename)
    }
    #[test]
    fn test_build_cc_result() {
        let _carbon_crush_result = build_cc_result(
            123.0,
            "myapp1",
            "main",
            "d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33",
            "http://whatever/job/123",
            1230.0,
            10.0,
        );
    }
}


src/scaph_reader.rs
use serde::Deserialize;
use serde::Serialize;
use serde_json::Value;
use std::fs::File;
use std::path::Path;
use std::path::PathBuf;

//use std::time::Duration;

pub type ScaphResults = Vec<Measure>;

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Measure {
    pub host: Host,
    pub consumers: Vec<Consumer>,
    pub sockets: Vec<Value>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Host {
    pub consumption: f32,
    pub timestamp: f64,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Consumer {
    pub exe: PathBuf,
    pub pid: i64,
    pub consumption: f32,
    pub timestamp: f64,
}

pub fn read_scaph_file(filename: &str) -> ScaphResults {
    //println!("reading scaph file {}", filename);
    let json_file_path = Path::new(filename);
    let file = File::open(json_file_path).expect("file not found");
    let results: ScaphResults = serde_json::from_reader(file).expect("error while reading");
    return results;
}

pub fn mean(data: Vec<f32>) -> Option<f32> {
    let sum = data.iter().sum::<f32>() as f32;
    let count = data.len();

    match count {
        positive if positive > 0 => Some(sum / count as f32),
        _ => None,
    }
}

pub fn average_consumption(scaphandre_file_name: &str, process_name: &str) -> f32 {
    println!(
        "Calculating average consumption of process[{}] from file[{}]",
        process_name, scaphandre_file_name
    );
    let scaph_results: ScaphResults = read_scaph_file(scaphandre_file_name);
    let mut consumptions: Vec<f32> = Vec::new();

    for meas in scaph_results {
        for proc in meas.consumers {
            if proc.exe.ends_with(process_name) {
                consumptions.push(proc.consumption)
            }
        }
    }
    match mean(consumptions) {
        Some(res) => res,
        None => panic!("Cannot calculate mean consumption"),
    }
}

pub fn process_duration_seconds(scaphandre_file_name: &str, process_name: &str) -> f32 {
    println!(
        "Extracting duration consumption of process: {} from file[{}]",
        process_name, scaphandre_file_name
    );
    let mut first_timestamp: f64 = 0.0;
    let mut last_timestamp: f64 = 0.0;

    let scaph_results: ScaphResults = read_scaph_file(scaphandre_file_name);

    for meas in scaph_results {
        for proc in meas.consumers {
            if proc.exe == PathBuf::from(process_name) {
                if first_timestamp == 0.0 {
                    first_timestamp = proc.timestamp;
                }
                last_timestamp = proc.timestamp;
            }
        }
    }

    (last_timestamp - first_timestamp) as f32
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test_reading_scaphandre_full_report() {
        let res: ScaphResults = read_scaph_file("./tests/scaphandre-full-report.json");
        assert_eq!(res.len(), 22);
    }
    #[test]
    fn test_reading_a_proc_name() {
        let res: ScaphResults = read_scaph_file("./tests/scaphandre-full-report.json");
        assert_eq!(res[20].consumers[1].exe, PathBuf::from("gnome-shell"));
    }
    #[test]
    fn test_reading_a_ts() {
        let res: ScaphResults = read_scaph_file("./tests/scaphandre-full-report.json");
        assert_eq!(res[20].consumers[1].timestamp, 1646408484.4658008 as f64);
    }

    #[test]
    fn test_mean() {
        let data = vec![1.0, 2.0, 3.0];
        let result = mean(data);
        match result {
            Some(res) => assert_eq!(res, 2.0),
            None => panic!(),
        }

        let data = vec![1.0, 2.0, 3.0, 4.0];
        let result = mean(data);
        match result {
            Some(res) => assert_eq!(res, 2.5),
            None => panic!(),
        }

        let empty: Vec<f32> = Vec::new();
        let result = mean(empty);
        match result {
            Some(res) => assert_eq!(res, 2.0),
            None => assert!(true),
        }
    }

    #[test]
    fn test_average_consumption() {
        let filename = "./tests/scaphandre-full-report.json";
        let process = "stress-ng";

        let res: f32 = average_consumption(filename, process);
        assert_eq!(res, 7269277.5 as f32);
    }

    #[test]
    fn test_average_consumption_small() {
        let filename = "./tests/scaphandre-simple-report.json";
        let process = "stress-ng";

        let res: f32 = average_consumption(filename, process);
        assert_eq!(res, 7867854.0 as f32);
    }

    #[test]
    fn test_process_duration_small() {
        let filename = "./tests/scaphandre-simple-report.json";
        let process = "stress-ng";

        let duration_seconds_f64 = process_duration_seconds(filename, process);

        assert_eq!(duration_seconds_f64, 2.0367724895477295 as f32);
    }
}


src/scaphandre_reader.rs
use serde::Deserialize;
use serde::Serialize;
use serde_json::Value;
use std::fs::File;
use std::path::PathBuf;

/// A vector of all Scaphandre measures
///
/// This is the equivalent of Scaphandre json exporter output.
pub type ScaphandreMeasures = Vec<Measure>;

/// A Scaphandre data point (measure).
///
#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Measure {
    pub host: Host,
    pub consumers: Vec<Consumer>,
    pub sockets: Vec<Value>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Host {
    /// Power usage of the host (in microWatts)
    pub consumption: f32,
    /// Timestamp of the measure
    pub timestamp: f64,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Consumer {
    pub exe: PathBuf,
    pub pid: i64,
    /// Power usage of the process (in microWatts)
    pub consumption: f32,
    /// Timestamp of the measure
    pub timestamp: f64,
}

/// Extracts Scaphandre measures from a Scaphandre json file
///
pub fn read_scaph_file(scaphandre_json_file: &PathBuf) -> ScaphandreMeasures {
    //dbg!("Reading scaphandre file {:?}", scaphandre_json_file);
    let file = File::open(scaphandre_json_file).expect("file not found");
    let results: ScaphandreMeasures = serde_json::from_reader(file).expect("error while reading");
    results
}

/// Returns the average (arithmetic mean) value of all elements of a vector
fn average(data: Vec<f32>) -> Option<f32> {
    let sum = data.iter().sum::<f32>() as f32;
    let count = data.len();

    match count {
        positive if positive > 0 => Some(sum / count as f32),
        _ => None,
    }
}
/// Calculate the average consumption of a given process (in micro Watts)
///
/// * Process are filtered on name
///
pub fn average_consumption_microwatt(scaphandre_json_file: &PathBuf, process_name: &str) -> f32 {
    println!(
        "Calculating average consumption of process[{}] from file[{:?}]",
        process_name, scaphandre_json_file
    );
    let scaph_results: ScaphandreMeasures = read_scaph_file(scaphandre_json_file);
    let mut consumptions: Vec<f32> = Vec::new();

    for meas in scaph_results {
        for proc in meas.consumers {
            if proc.exe.ends_with(process_name) {
                consumptions.push(proc.consumption)
            }
        }
    }
    match average(consumptions) {
        Some(res) => res,
        None => panic!("Cannot calculate average consumption"),
    }
}

/// Extract the total duration of a process (in seconds)
///
/// * Duration is obtained by reading the scaphandre json output, filtering on process name,
/// and calculating the difference between the last and first timestamps of the process.
/// * returns 0 if process cannot be found
///
pub fn process_duration_seconds(scaphandre_json_file: &PathBuf, process_name: &str) -> f32 {
    println!(
        "Extracting duration consumption of process: {} from file[{:?}]",
        process_name, scaphandre_json_file
    );
    let mut first_timestamp: f64 = 0.0;
    let mut last_timestamp: f64 = 0.0;

    let scaph_results: ScaphandreMeasures = read_scaph_file(scaphandre_json_file);

    for measure in scaph_results {
        for process in measure.consumers {
            if process.exe == PathBuf::from(process_name) {
                if first_timestamp == 0.0 {
                    first_timestamp = process.timestamp;
                }
                last_timestamp = process.timestamp;
            }
        }
    }
    // FIXME: handle the case where process can never be found
    // FIXME: what happens the process appears only once ?
    (last_timestamp - first_timestamp) as f32
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_reading_scaphandre_full_report() {
        let scaphandre_json_file = PathBuf::from("./tests/scaphandre-full-report.json");
        let measures: ScaphandreMeasures = read_scaph_file(&scaphandre_json_file);
        assert_eq!(measures.len(), 22);
    }
    #[test]
    fn test_reading_a_proc_name() {
        let scaphandre_json_file = PathBuf::from("./tests/scaphandre-full-report.json");
        let measures: ScaphandreMeasures = read_scaph_file(&scaphandre_json_file);
        assert_eq!(measures[20].consumers[1].exe, PathBuf::from("gnome-shell"));
    }
    #[test]
    fn test_reading_a_ts() {
        let scaphandre_json_file = PathBuf::from("./tests/scaphandre-full-report.json");
        let measures: ScaphandreMeasures = read_scaph_file(&scaphandre_json_file);
        assert_eq!(
            measures[20].consumers[1].timestamp,
            1646408484.4658008 as f64
        );
    }

    #[test]
    fn test_mean() {
        let data = vec![1.0, 2.0, 3.0];
        let result = average(data);
        match result {
            Some(res) => assert_eq!(res, 2.0),
            None => panic!(),
        }

        let data = vec![1.0, 2.0, 3.0, 4.0];
        let result = average(data);
        match result {
            Some(res) => assert_eq!(res, 2.5),
            None => panic!(),
        }

        let empty: Vec<f32> = Vec::new();
        let result = average(empty);
        match result {
            Some(res) => assert_eq!(res, 2.0),
            None => assert!(true),
        }
    }

    #[test]
    fn test_average_consumption_full() {
        let scaphandre_json_file = PathBuf::from("./tests/scaphandre-full-report.json");
        let process_name = "stress-ng";

        let res: f32 = average_consumption_microwatt(&scaphandre_json_file, process_name);
        assert_eq!(res, 7269277.5 as f32);
    }

    #[test]
    fn test_average_consumption_simple() {
        let scaphandre_json_file = PathBuf::from("./tests/scaphandre-simple-report.json");
        let process_name = "stress-ng";

        let res: f32 = average_consumption_microwatt(&scaphandre_json_file, process_name);
        assert_eq!(res, 7867854.0 as f32);
    }

    #[test]
    fn test_process_duration_simple() {
        let scaphandre_json_file = PathBuf::from("./tests/scaphandre-simple-report.json");
        let process_name = "stress-ng";

        let duration_seconds_f64 = process_duration_seconds(&scaphandre_json_file, process_name);

        assert_eq!(duration_seconds_f64, 2.0367724895477295 as f32);
    }
}


src/carboncrush_exporter.rs
use std::fmt;
use std::path::PathBuf;

use serde::Deserialize;
use serde::Serialize;

use quick_junit::*;

/// The structure that describe Carbon Crush input
#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CarbonCrushResult {
    /// Power (uW)
    pub consumption: String,
    pub app_id: String,
    // duration in seconds
    pub duration: String,
    pub branch: String,
    pub commit_sha: String,
    /// Energy used (micro Watts-second)
    pub energy: String,
    /// Energy used (Wh)
    pub energy_watthours: f32,
    pub ci_pipeline_url: String,
}

impl fmt::Display for CarbonCrushResult {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "consumption: {}", self.consumption)
    }
}

/// Save a carbon crush results as a file
pub fn save_carboncrush_file(
    carbon_crush_result: &CarbonCrushResult,
    carboncrush_json_file: PathBuf,
) {
    println!(
        "Saving results: {:?} to {:?}",
        carbon_crush_result, carboncrush_json_file
    );
    // Save the JSON structure into the other file.
    std::fs::write(
        carboncrush_json_file,
        serde_json::to_string_pretty(carbon_crush_result).unwrap(),
    )
    .unwrap();
}

pub fn save_as_junit_report(carbon_crush_result: &CarbonCrushResult, report_file: PathBuf) {
    let report = build_junit_report(carbon_crush_result);
    // Save the junit report into file.
    std::fs::write(report_file, report.to_string().unwrap()).unwrap();
}

/// Build a carbon crush data structure with the passed values.
///
pub fn build_carboncrush_result(
    consumption_uw: f32,
    app_id: &str,
    branch: &str,
    commit_sha: &str,
    ci_pipeline_url: &str,
    duration_seconds: f32,
) -> CarbonCrushResult {
    let energy_microwattseconds: f32 = get_energy_uws(consumption_uw, duration_seconds);
    let energy_watthours: f32 = get_energy_wh(consumption_uw, duration_seconds);

    CarbonCrushResult {
        consumption: consumption_uw.to_string(),
        app_id: app_id.to_string(),
        energy: energy_microwattseconds.to_string(),
        branch: branch.to_string(),
        ci_pipeline_url: ci_pipeline_url.to_string(),
        commit_sha: commit_sha.to_string(),
        duration: duration_seconds.to_string(),
        energy_watthours,
    }
}

pub fn build_junit_report(carbon_crush_result: &CarbonCrushResult) -> Report {
    let mut report = Report::new("Power measure");
    let test_suite_name = carbon_crush_result.app_id.as_str();
    let mut test_suite = TestSuite::new(test_suite_name);
    let mut success_case = TestCase::new(
        carbon_crush_result.commit_sha.as_str(),
        TestCaseStatus::success(),
    );

    success_case.set_system_out(format!("{:?}", carbon_crush_result));

    test_suite.add_test_cases([success_case]);
    report.add_test_suite(test_suite);
    report
}

/// Energy in microWatt-seconds (given a power in uW and duration in seconds)
pub fn get_energy_uws(power_microwatts: f32, duration_seconds: f32) -> f32 {
    power_microwatts * duration_seconds
}

/// Energy in Wh (given a power in uW and duration in seconds)
pub fn get_energy_wh(power_microwatts: f32, duration_seconds: f32) -> f32 {
    let consumption_watts = convert_microwatts_to_watts(power_microwatts);
    let duration_hours = duration_seconds / 3600_f32;

    consumption_watts * duration_hours
}

pub fn convert_microwatts_to_watts(micro_watts: f32) -> f32 {
    micro_watts * 1e-6_f32
}

#[test]
fn test_convert() {
    let power_microwatts = 1000000_f32;
    let duration_seconds = 3600_f32;
    let e_wh = get_energy_wh(power_microwatts, duration_seconds);
    assert_eq!(e_wh, 1_f32);

    let r_wh = convert_microwatts_to_watts(1000000_f32);
    assert_eq!(r_wh, 1_f32);
}

#[test]
fn test_build_cc_result() {
    let _carboncrush_result = build_carboncrush_result(
        123.0,
        "myapp1",
        "main",
        "d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33",
        "http://whatever/job/123",
        1230.0,
    );
}
#[test]
fn test_save_cc_file() {
    let carboncrush_json_file = PathBuf::from("tests/test-generated-result.json");
    let carboncrush_result = build_carboncrush_result(
        123.0,
        "myapp1",
        "main",
        "d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33",
        "http://whatever/job/123",
        1230.0,
    );
    save_carboncrush_file(&carboncrush_result, carboncrush_json_file)
}
#[test]
fn test_build_junit_report() {
    let carboncrush_result = build_carboncrush_result(
        123.0,
        "myapp1",
        "main",
        "d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33",
        "http://whatever/job/123",
        10_f32,
    );
    let report: Report = build_junit_report(&carboncrush_result);
    println!("{}", report.to_string().unwrap());
    const EXPECTED_XML: &str = r#"<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="Power measure" tests="1" failures="0" errors="0">
    <testsuite name="myapp1" tests="1" disabled="0" errors="0" failures="0">
        <testcase name="d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33">
            <system-out>CarbonCrushResult { consumption: &quot;123&quot;, app_id: &quot;myapp1&quot;, duration: &quot;10&quot;, branch: &quot;main&quot;, commit_sha: &quot;d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33&quot;, energy: &quot;1230&quot;, energy_watthours: 3.416667e-7, ci_pipeline_url: &quot;http://whatever/job/123&quot; }</system-out>
        </testcase>
    </testsuite>
</testsuites>
"#;
    assert_eq!(report.to_string().unwrap(), EXPECTED_XML);
}
#[test]
fn test_save_junit_report() {
    let carboncrush_result = build_carboncrush_result(
        123.0,
        "myapp1",
        "main",
        "d50e3b5ed5c27a848008abd5beb3d9e6c37c3f33",
        "http://whatever/job/123",
        1230.0,
    );
    let mut junit_report: PathBuf = PathBuf::new();
    junit_report.push("tests/carbon-crush-report.xml");
    save_as_junit_report(&carboncrush_result, junit_report);
}
